src/main/scala/utility/LogPerfHelper.scala

```diff
@@ -37,3 +38,10 @@ class LogPerfHelper extends BlackBox with HasBlackBoxInline {
       |""".stripMargin
   setInline("LogPerfHelper.v", verilog)
 }
+
+object LogPerfControl {
+  private val instances = scala.collection.mutable.ListBuffer.empty[LogPerfIO]
+  private def instantiate(): LogPerfIO = instances.addOne(WireInit(Module(new LogPerfHelper).io)).last
+
+  def apply(): LogPerfIO = instances.find(gen => isVisible(gen)).getOrElse(instantiate())
```

Comment:
One question: after this change, the design will not be compatible with older versions of Chisel before v6.1.0. Is this accpepted by XiangShan and its submodules? @Tang-Haojin 

Instead of calling `isVisible` here, I would prefer using Scala reflection to call the LogPerfControl in DiffTest. Only if that does not exist (the module using utility is not simulated by DiffTest), we use the one currently defined in utility. Then we don't need to copy the isVisible code here and the code would always align with DiffTest. I believe in the future, we will replace the one in DiffTest with `tapAndRead`.

Reply:
But `tapAndRead` also has some dedup issues so I write a transform to dedup manually.

Reply:
Previously I also try to use tapAndRead for perfIO(Register source first and assign tap to sink somewhere). However, It brings a lot of middle ports like addSource and addSink.

I think that's because chisel use probe ports (can be seen in firrtl module ports) to implement tap. And output probe(child source to parent sink) can be replaced by hierarchical reference like aa.bb.cc, but input probe will still be generated as IOs. 

Does current tap supports hidden middle IO？I also agree modified tap is better to logPerf.

Reply:
No, chisel still cannot generate non-hierarchical XMR in Verilog, and it generates IO ports. But I think it is okay.

Reply:
How many bore ports generate by tap LogPerf? It seems currently tapAndRead will generate bore ports for every LogPerf, and it will generate too many reduplicated xx_bore_yy in Top. Maybe we can firstly use isVisible to reduce LogPerf sinks (from thousands to hundreds) and then use tap to assign these sinks.

Another question is some Verification scene requires no additional ports added to DUT( same to final release version without LogPerf or Difftest). That's why we support optional Hierarchical wiring for Difftest. Maybe we can still provide an optional Helper, but use tap by default?

Reply:
`tapAndRead` does have dedup problem so I wrote a transform to handle it manually. I agree with preserving Helper as an option.

Reply:
> No, chisel still cannot generate non-hierarchical XMR in Verilog, and it generates IO ports. But I think it is okay.

Is this behavior for only "tap from a child module to a parent module"? Or, does it apply to "tap a signal in a child module"?

If the latter does generate XMR, maybe we can consider moving all printf's for performance counters to the top. The only thing we need to do is tapAndRead these counters in the top. Then we only need one top-level Help.

Reply:
That is ok but printing hierarchical name (%m) may be a new problem.

Reply:
> Seems name can also be registered to list or extracted from Variable? Maybe we can also consider LogPerfHelper as Difftest, apply somewhere and collect in SimTop. So that all perf print's is contained in a single module, we can easily remove it if necessary.
> 
> The solution may also deal with some generated probe (LogPerf, ChiselDB ...) out of param control. So all debug signals can be seperate from DUT, to better meets verification demand.

Is hierarchical name also easy to be extracted? Like "TOP.SimTop.l_soc.core_with_l2.core.memBlock.dcache.dcache.probeQueue.entries_4". I think there must be a solution because the hierarchical XMR actually contains such information.

Reply:
For this PR, it is okay to be merged for me, but may check the compatibility of submodules first.

Reply:
I can have a try for hierarchical name first. The problem is instance can only be accessed after elaboration.

If it works, maybe we can collect logPerf directly.

Reply:
> That is ok but printing hierarchical name (%m) may be a new problem.

I see. This is indeed a problem.

Reply:
Hierarchical name(for both data or module) can be accessed by xx.pathName after circuit elaboration ( Stage.execute), which means we may need another method called after elaboration, generating macros of hierarchical module name, is it acceptable?
Alternatively, during elaboration, we can get currentModule name up to top module name using recursive DataMirror.getParent. Like SimTop.Mod1.Mod2... instead of SimTop.Inst1.Inst2... But it may be mixed up with multi instances with same modName

Reply:
Seems ChiselDB and Constantin already have similar calling interface after elaboration like `addToFileRegisters`

Reply:
> Hierarchical name(for both data or module) can be accessed by xx.pathName after circuit elaboration ( Stage.execute), which means we may need another method called after elaboration, generating macros of hierarchical module name, is it acceptable? Alternatively, during elaboration, we can get currentModule name up to top module name using recursive DataMirror.getParent. Like SimTop.Mod1.Mod2... instead of SimTop.Inst1.Inst2... But it may be mixed up with multi instances with same modName

I think it's good. We can add this optional interface to difftest or utility.